sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java

```diff
@@ -0,0 +1,743 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.trace;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.EvictingQueue;
+import com.google.protobuf.BoolValue;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Timestamp;
+import com.google.protobuf.UInt32Value;
+import io.opentelemetry.proto.trace.v1.Span.Attributes;
+import io.opentelemetry.proto.trace.v1.Span.Links;
+import io.opentelemetry.proto.trace.v1.Span.SpanKind;
+import io.opentelemetry.proto.trace.v1.Span.TimedEvents;
+import io.opentelemetry.proto.trace.v1.TruncatableString;
+import io.opentelemetry.resource.Resource;
+import io.opentelemetry.sdk.internal.Clock;
+import io.opentelemetry.sdk.internal.TimestampConverter;
+import io.opentelemetry.sdk.trace.config.TraceConfig;
+import io.opentelemetry.sdk.trace.internal.ConcurrentIntrusiveList.Element;
+import io.opentelemetry.trace.AttributeValue;
+import io.opentelemetry.trace.Event;
+import io.opentelemetry.trace.Link;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.SpanContext;
+import io.opentelemetry.trace.SpanData;
+import io.opentelemetry.trace.SpanData.TimedEvent;
+import io.opentelemetry.trace.SpanId;
+import io.opentelemetry.trace.Status;
+import io.opentelemetry.trace.TraceId;
+import io.opentelemetry.trace.Tracer;
+import io.opentelemetry.trace.Tracestate;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+/** Implementation for the {@link Span} class that records trace events. */
+@ThreadSafe
+public final class RecordEventsSpanImpl implements SpanSdk, Element<RecordEventsSpanImpl> {
+  private static final Logger logger = Logger.getLogger(Tracer.class.getName());
+
+  // Contains the identifiers associated with this Span.
+  private final SpanContext context;
+  // The parent SpanId of this span. Null if this is a root span.
+  @Nullable private final SpanId parentSpanId;
+  // True if the parent is on a different process.
+  @Nullable private final Boolean hasRemoteParent;
```

Comment:
This has been removed.

Reply:
It's still being used in the proto: https://github.com/open-telemetry/opentelemetry-java/blob/ca0ec9dfa83fd8c4b1a9e4a5af3db75cf64c6cbc/sdk/src/main/proto/trace/v1/trace.proto#L240-L245.